### DISCUSSION TITLE
internal agent for GetSimilarArtists (in addition to last.fm agent)

### DISCUSSION THREAD

**Discussion by metalheim** [navidrome/navidrome#3508] [Ideas]
It would be nice if the data for the GetSimilarArtists endpoint could be queried internally.
This should be an *option* (switchable between internal and lasftm), maybe defaulting to internal if you don't have lastfm enabled.

### Pros:
1. performance (last.fm api isn't always very performant)
2. no reliancy on external service (some people dont wan't last.fm to be queried)
3. (related to 2.) Solely tag-based. You can get similar artists that even are not in lasf.tm/musicbrainz. Everything will follow the tags you set

### Cons:
Solely tag-based. Some artist "relations" cannot be stored in tags (such as "2 bands toured together" or "1 artist is the brother of another", "2 bands guitar riffs have a similar style/sound"). 
These relations will not be accounted for when using the "internal" agent. I'm not even sure how comprehensive last.fm data is and if its manually curated in some way

## Ideas how to determine "similarity"
1. Mutual song appearances. If an artist is on a song of another artist, it is somewhat similar.
2. Mutual album appearances. If an artist appears on an album of another artist, it is somewhat similar
4. Mutual genres (i have no immediate good idea how to solve this, probably would need to roll up genres to artists to be performant? Then we could compare tuples of genres from one artist to all others)

Points 1. and 2. are fairly easy to do with the new (after #2709 ) database layout.
The beauty of the new DB design is that it's easy to not only consider track appearances ("featurings"), but also data form "involvedpeople" (`TIPL`), Composer, Producer etc.
Here's a simple PoC sql query that finds similar artists to artistId `22gKASSTVJIX1d2FAIRIcg` based on mutual album and track appearances:

``` sql
SELECT smar.*, artist.name
FROM (
	SELECT COUNT(mf.media_file_id) as similarity, mfa_ra.artist_id FROM 
		(SELECT * FROM media_file_artists 
		WHERE artist_id = '22gKASSTVJIX1d2FAIRIcg') mf
	JOIN media_file_artists mfa_ra
		ON mf.media_file_id = mfa_ra.media_file_id
	GROUP BY mfa_ra.artist_id 
	ORDER BY similarity DESC
) as smar
JOIN artist 
ON smar.artist_id = artist.id
WHERE artist_id <> '22gKASSTVJIX1d2FAIRIcg'
LIMIT 10
```

artistId `22gKASSTVJIX1d2FAIRIcg` is 'Eric Clapton' in my library.
Output of above script is:

63	7kkn5CXckLNP85vTjVdt1L	B.B. King
31	72mR8zKMOGKIthTlTw51t1	The Bluesbreakers
31	6eCl0t8ia6LQDLcB1XmkyT	John Mayall & The Bluesbreakers
31	5yzacOj5mFHoOcGVVmw3SL	John Mayal
30	62u9W2H8SD5FEAe4ETubRk	Robert Johnson
11	0PRJ5WgdzG2pk2ZibJ3Qyp	Van Morrison
10	6Ie24AfrVlMIlbM8rIhqmA	Traditional
9	5ySdj8HiyCrY3E1XGH6Bef	Cream
8	0yxRWmZM9QDVTwh9v2mjTA	J.J. Cale
6	7yCBTLmUnUWCC5etBqqh2s	Jerry Lynn Williams

**Comment by deluan** [comment]
When I implemented agents back in 2021, I had this same idea: to implement all features that requires Last.fm using local data. I even implemented the simplest one, [GetArtistTopSongs](https://github.com/navidrome/navidrome/blob/93adda66d96348f618052a16818d032ffd1212f9/core/agents/local_agent.go#L24) in a "local" agent (it works, if you put "local" in your Agents list).

We can definitely expand this local agent with more functionality, and the [agent interface](https://github.com/navidrome/navidrome/blob/93adda66d96348f618052a16818d032ffd1212f9/core/agents/README.md) will be one of the first to be available for plugins. We could also implement SimilarSongs by using #2259, for example. 

IIRC, Last.fm uses the scrobble data it collects to create a graph of similar artits/albums/songs, which makes sense if you have volume (as they do)

Anyway, your PoC make sense but it requires lots of music to be effective. I tried a few more isolated artists and it didn't return anything, as expected.

Using genres could be a better approach, but I decided to drop artists genres from the DB as in this multi-artist DB, we could endup with lots of wrong artists genres. Because of remixes, we could get things very wrong, like [_Ramones_](https://www.youtube.com/watch?v=5aNc5aEMjhs) appearing under Reggae. 

Not sure what would be a next step here. Let's revisit this once I have more time (after the BFR release).

**Reply by metalheim** [reply]
I agree that my very simple approach only works with lots of music (and navidrome insights show that most people don't a huge library).
Without additional relationships (that a pure tag-based approach cannot deliver), my idea probably remains a specialty and most people are better off with last.fm/spotify/listenbrainz/whatever.

But i'm willing to invest some time writing a local agent as a plugin when available :)

**Comment by metalheim** [comment]
I am currently working on a plugin that implements a local Agent and will be able to resolve (fully local):
- getSimilarArtists
- getTopSongs
- getSimilarSongs (once that becomes available from nd side)

I'm also investigating in parsing a `.nfo` (kodi-style metadata sidecar file). This would also enable Agent functionalities:
- get_artist_biography
- nd_get_album_info (Description)

Will update here once development has progressed a bit and the new plugin system is merged to navidromes main branch.
